proposals/0000-updated-integers.md

```diff
@@ -0,0 +1,615 @@
+  func divided(by rhs: Self) -> Self
+
+  mutating func add( rhs: Self)
+  mutating func subtract( rhs: Self)
+  mutating func multiply(by rhs: Self)
+  mutating func divide(by rhs: Self)
+}
+```
+
+#### `SignedArithmetic`
+
+The `SignedArithmetic` is for the numbers that can be negated.
+ 
+```Swift
+public protocol SignedArithmetic : Arithmetic {
+  func negate() -> Self
+}
+```
+
+#### `Integer`
+
+The `Integer` protocol is the basis for all the integer types provided by the
+standard library.
+
+The `isEqual(to:)` and `isLess(than:)` methods provide implementations for the
+`Equatable` and `Comparable` protocol conformances. Similar to how arithmetic
+operations are dispatched in `Arithmetic`, `==` and `<` operators for
+homogeneous comparisons are implemented as generic free functions invoking
+`isEqual(to:)` and `isLess(than:)` protocol methods respectively.
+
+This protocol adds 3 new initializers to the parameterless one, inherited from
+`Arithmetic`. These initializers allow to construct values of type from
+instances of any other type, conforming to `Integer`, using different
+strategies:
+
+  - Perform checks whether the value is representable in `Self`
+  
+  - Try to represent the value without bounds checks
+  
+  - Substitute values beyond `Self` bounds with maximum/minimum value of
+    `Self` respectively
+
+The `Absolutevalue` associated type refers to the type that is able to hold an
+absolute value of any possible value of `Self`.
+
+Concrete types do not have to provide a typealias for it as it can be inferred
+from the `absoluteValue` property. This property can be useful in operations
+that are simpler to implement in terms of unsigned values, for example,
+printing a value of an integer, which is just printing a '-' character in front
+of an absolute value.
+
+Please note, that `absoluteValue` property *should not* be preferred to the
+`abs` function, who's return value is of the same type as its argument.
+
+```Swift
+public protocol Integer:
+  Comparable, Arithmetic,
+  IntegerLiteralConvertible, CustomStringConvertible {
+
+  associatedtype AbsoluteValue : Integer // this is not the actual code
+
+  static var isSigned: Bool { get }
+
+  var absoluteValue: AbsoluteValue { get }
+
+  func isEqual(to rhs: Self) -> Bool
+  func isLess(than rhs: Self) -> Bool
+
+  /// Creates an instance of `Self` from `source` if it is representable.
+  ///
+  /// - Precondition: the value of `source` is representable in `Self`.
+  init<T : Integer>(_ source: T)
+
+  /// Creates in instance of `Self` from `source` by sign-extending it
+  /// indefinitely and then truncating to fit `Self`.
+  init<T : Integer>(extendingOrTruncating source: T)
+
+  /// Creates in instance of `Self` containing the closest representable
+  /// value of `source`.
+  init<T : Integer>(clamping source: T)
+
+  /// Returns n-th word, counting from the right, of the underlying
+  /// representation of `self`.
+  func nthWord(n: Int) -> UInt
```

Comment:
Is there a way to know how many words there are to get, other than `(bitWidth+sizeof(UInt)-1) / sizeof(UInt)`?